File: app/src/main/java/knf/kuma/tv/TVServersFactory.java
package knf.kuma.tv;

import android.app.Activity;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.v17.leanback.widget.Presenter;
import android.util.Log;

import com.afollestad.materialdialogs.MaterialDialog;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.CustomEvent;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import knf.kuma.commons.BypassUtil;
import knf.kuma.database.CacheDB;
import knf.kuma.pojos.AnimeObject;
import knf.kuma.pojos.DownloadObject;
import knf.kuma.pojos.RecordObject;
import knf.kuma.tv.exoplayer.TVPlayer;
import knf.kuma.tv.streaming.TVServerSelection;
import knf.kuma.tv.streaming.TVServerSelectionFragment;
import knf.kuma.videoservers.Option;
import knf.kuma.videoservers.Server;
import knf.kuma.videoservers.VideoServer;
import xdroid.toaster.Toaster;

public class TVServersFactory {
    public static int REQUEST_CODE_LIST = 4456;
    public static int REQUEST_CODE_OPTION = 6157;

    private Activity activity;
    private String url;
    private AnimeObject.WebInfo.AnimeChapter chapter;
    private DownloadObject downloadObject;
    private ServersInterface serversInterface;

    private List<Server> servers = new ArrayList<>();

    private VideoServer current;

    private Presenter.ViewHolder viewHolder;

    private TVServersFactory(Activity activity, String url, AnimeObject.WebInfo.AnimeChapter chapter, Presenter.ViewHolder viewHolder, ServersInterface serversInterface) {
        this.activity = activity;
        this.url = url;
        this.chapter = chapter;
        this.downloadObject = DownloadObject.fromChapter(chapter, false);
        this.viewHolder = viewHolder;
        this.serversInterface = serversInterface;
    }

    public static void start(final Activity activity, final String url, final AnimeObject.WebInfo.AnimeChapter chapter, final ServersInterface serversInterface) {
        start(activity, url, chapter, null, serversInterface);
    }

    public static void start(final Activity activity, final String url, final AnimeObject.WebInfo.AnimeChapter chapter, Presenter.ViewHolder viewHolder, final ServersInterface serversInterface) {
        final MaterialDialog dialog = new MaterialDialog.Builder(activity)
                .content("Obteniendo servidores")
                .progress(true, 0)
                .build();
        dialog.show();
        AsyncTask.execute(() -> {
            TVServersFactory factory = new TVServersFactory(activity, url, chapter, viewHolder, serversInterface);
            serversInterface.onReady(factory);
            factory.get(dialog);
        });
    }

    public void showServerList() {
        new Handler(Looper.getMainLooper()).post(() -> {
            try {
                if (servers.size() == 0) {
                    Toaster.toast("Sin servidores disponibles");
                    serversInterface.onFinish(false, false);
                } else {
                    Bundle bundle = new Bundle();
                    bundle.putStringArrayList(TVServerSelectionFragment.SERVERS_DATA, (ArrayList<String>) Server.getNames(servers));
                    activity.startActivityForResult(new Intent(activity, TVServerSelection.class)
                            .putExtras(bundle), REQUEST_CODE_LIST);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        });

    }

    public void analizeServer(int position) {
        final MaterialDialog dialog = new MaterialDialog.Builder(activity)
                .content("Obteniendo link")
                .progress(true, 0)
                .build();
        dialog.show();
        AsyncTask.execute(() -> {
            try {
                String text = servers.get(position).getName();
                final VideoServer server = servers.get(position).getVideoServer();
                dialog.dismiss();
                if (server == null && servers.size() == 1) {
                    Toaster.toast("Error en servidor, intente mas tarde");
                    serversInterface.onFinish(false, false);
                } else if (server == null) {
                    Toaster.toast("Error en servidor");
                    showServerList();
                } else if (server.options.size() == 0) {
                    Toaster.toast("Error en servidor");
                    showServerList();
                } else if (server.haveOptions()) {
                    showOptions(server);
                } else {
                    switch (text.toLowerCase()) {
                        case "mega":
                            Toaster.toast("No se puede usar Mega en TV");
                            showServerList();
                            break;
                        default:
                            startStreaming(server.getOption());
                            break;
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
    }

    public void analizeOption(int position) {
        if (current != null)
            startStreaming(current.options.get(position));
    }

    private void showOptions(final VideoServer server) {
        this.current = server;
        Bundle bundle = new Bundle();
        bundle.putStringArrayList(TVServerSelectionFragment.VIDEO_DATA, (ArrayList<String>) Option.getNames(server.options));
        bundle.putString("name", server.name);
        activity.startActivityForResult(new Intent(activity, TVServerSelection.class)
                .putExtras(bundle), REQUEST_CODE_OPTION);
    }

    private void startStreaming(Option option) {
        CacheDB.INSTANCE.chaptersDAO().addChapter(chapter);
        CacheDB.INSTANCE.recordsDAO().add(RecordObject.fromChapter(chapter));
        Answers.getInstance().logCustom(new CustomEvent("Streaming").putCustomAttribute("Server", option.server));
        activity.startActivity(new Intent(activity, TVPlayer.class)
                .putExtra("url", option.url)
                .putExtra("title", downloadObject.name)
                .putExtra("chapter", downloadObject.chapter));
        serversInterface.onFinish(false, true);
    }

    public Presenter.ViewHolder getViewHolder() {
        return viewHolder;
    }

    public void get(MaterialDialog dialog) {
        try {
            Log.e("Url", url);
            Document main = Jsoup.connect(url).timeout(5000).cookies(BypassUtil.getMapCookie(activity)).userAgent(BypassUtil.userAgent).get();
            Elements descargas = main.select("table.RTbl.Dwnl").first().select("a.Button.Sm.fa-download");
            List<Server> servers = new ArrayList<>();
            for (Element e : descargas) {
                String z = e.attr("href");
                z = z.substring(z.lastIndexOf("http"));
                Server server = Server.check(activity, z);
                if (server != null)
                    servers.add(server);
            }
            Elements s_script = main.select("script");
            String j = "";
            for (Element element : s_script) {
                String s_el = element.outerHtml();
                if (s_el.contains("var video = [];")) {
                    j = s_el;
                    break;
                }
            }
            String[] parts = j.substring(j.indexOf("var video = [];") + 14, j.indexOf("$(document).ready(function()")).split("video\\[[^a-z]*\\]");
            for (String baseLink : parts) {
                Server server = Server.check(activity, baseLink);
                if (server != null)
                    servers.add(server);
            }
            Collections.sort(servers);
            this.servers = servers;
            dialog.dismiss();
            showServerList();
        } catch (Exception e) {
            e.printStackTrace();
            this.servers = new ArrayList<>();
            dialog.dismiss();
            serversInterface.onFinish(false, false);
        }
    }

    public interface ServersInterface {
        void onReady(TVServersFactory serversFactory);

        void onFinish(boolean started, boolean success);
    }
}

File: app/src/main/java/knf/kuma/tv/TVServersFactory.kt
package knf.kuma.tv

import android.content.Intent
import android.net.Uri
import androidx.fragment.app.FragmentActivity
import androidx.leanback.widget.Presenter
import androidx.lifecycle.lifecycleScope
import com.google.firebase.crashlytics.FirebaseCrashlytics
import knf.kuma.backup.firestore.syncData
import knf.kuma.commons.JsExtractor
import knf.kuma.commons.doOnUIGlobal
import knf.kuma.commons.iterator
import knf.kuma.database.CacheDB
import knf.kuma.player.openWebPlayer
import knf.kuma.pojos.AnimeObject
import knf.kuma.pojos.DownloadObject
import knf.kuma.pojos.RecordObject
import knf.kuma.pojos.SeenObject
import knf.kuma.pojos.av1.Chapter
import knf.kuma.pojos.av1.RecentAV1
import knf.kuma.pojos.av1.Record
import knf.kuma.tv.exoplayer.TVPlayer
import knf.kuma.tv.streaming.TVMultiSelection
import knf.kuma.tv.streaming.TVServerSelection
import knf.kuma.tv.streaming.TVServerSelectionFragment
import knf.kuma.videoservers.Option
import knf.kuma.videoservers.Server
import knf.kuma.videoservers.VideoServer
import knf.kuma.videoservers.WebServer
import kotlinx.coroutines.launch
import org.jetbrains.anko.doAsync
import org.json.JSONObject
import xdroid.toaster.Toaster
import java.util.Locale


class TVServersFactory private constructor(
    private val activity: FragmentActivity,
    private val url: String,
    private val name: String,
    private val chapter: String,
    private val record: Record,
    val viewHolder: Presenter.ViewHolder?,
    private val serversInterface: ServersInterface
) {
    private var servers: MutableList<Server> = ArrayList()
    private var subServes: List<Server> = emptyList()
    private var dubServers: List<Server> = emptyList()

    private var current: VideoServer? = null

    fun showServerList() {
        doOnUIGlobal {
            try {
                if (servers.isEmpty()) {
                    Toaster.toast("Sin servidores disponibles")
                    serversInterface.onFinish(false, false)
                } else {
                    activity.startActivityForResult(
                        Intent(activity, TVServerSelection::class.java)
                            .putExtra(TVServerSelectionFragment.SERVERS_DATA, Server.getNames(servers) as ArrayList),
                        REQUEST_CODE_LIST
                    )
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

    }

    fun analyzeMulti(position: Int) {
        when(position) {
            0 -> this.servers = subServes.toMutableList()
            1 -> this.servers = dubServers.toMutableList()
        }
        showServerList()
    }

    fun analyzeServer(position: Int) {
        doAsync {
            try {
                val text = servers[position].name
                val server = servers[position].verified
                if (server == null && servers.size == 1) {
                    Toaster.toast("Error en servidor, intente mas tarde")
                    serversInterface.onFinish(false, false)
                } else if (server == null) {
                    Toaster.toast("Error en servidor")
                    showServerList()
                } else if (server.options.size == 0) {
                    Toaster.toast("Error en servidor")
                    showServerList()
                } else if (server.haveOptions()) {
                    showOptions(server)
                } else if (servers[position] is WebServer) {
                    try {
                        openWebPlayer(activity, server.option.url!!, name)
                        doAsync {
                            CacheDB.INSTANCE.recordAV1DAO().addChapter(record)
                            syncData {
                                history()
                            }
                        }
                        serversInterface.onFinish(false, true)
                    } catch (_: Exception) {
                        Toaster.toast("Error al abrir explorador web")
                        showServerList()
                    }

                } else {
                    val serverName = text.lowercase(Locale.getDefault())
                    when {
                         serverName.contains("mega") -> {
                            Toaster.toast("No se puede usar Mega en TV")
                            showServerList()
                        }
                        else -> startStreaming(server.option)
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun analyzeOption(position: Int) {
        current?.let { startStreaming(it.options[position]) }
    }

    private fun showOptions(server: VideoServer) {
        this.current = server
        activity.startActivityForResult(
            Intent(activity, TVServerSelection::class.java)
                .putExtra("name", server.name)
                .putExtra(
                    TVServerSelectionFragment.VIDEO_DATA, (server.options.map { it.name ?: "" } as? ArrayList)
                        ?: arrayListOf<String>()
                ),
            REQUEST_CODE_OPTION
        )
    }

    private fun startStreaming(option: Option) {
        doAsync {
            CacheDB.INSTANCE.recordAV1DAO().addChapter(record)
            syncData {
                history()
            }
        }
        activity.startActivity(Intent(activity, TVPlayer::class.java).apply {
            setDataAndType(Uri.parse(option.url), "video/*")
            putExtra("title", name)
            putExtra("chapter", chapter)
            putStringArrayListExtra("headers", ArrayList(option.headers?.createHeadersList()?: emptyList()))
        })
        serversInterface.onFinish(false, true)
    }

    fun get() {
        activity.lifecycleScope.launch {
            try {
                val response = JsExtractor.processLinkMultiple(url, listOf("embeds", "downloads"))
                val subServers = mutableListOf<Server>()
                val dubServers = mutableListOf<Server>()
                response.forEach { (_, jSONArray) ->
                    jSONArray?.getJSONObject(0)?.let {
                        if (it.has("SUB")) {
                            for (sub in it.getJSONArray("SUB")) {
                                val name = sub.getString("server")
                                val url = sub.getString("url")
                                if (name == "MP4Upload") {
                                    if (subServers.find { it.baseLink.contains(url.substringAfterLast("/")) } != null){
                                        continue
                                    }
                                }
                                if (name == "PDrain") {
                                    if (subServers.find { it.baseLink.substringBeforeLast("?").contains(url.substringBeforeLast("?")) } != null){
                                        continue
                                    }
                                }
                                if (name == "1Fichier") {
                                    continue
                                }
                                val server = Server.check(activity, url)
                                if (subServers.find { it.baseLink.substringAfterLast("/") == url.substringAfterLast("/") } == null) {
                                    subServers.add(server?: WebServer(activity, url, name))
                                }
                            }
                        }
                        if (it.has("DUB")) {
                            for (dub in it.getJSONArray("DUB")) {
                                val name = dub.getString("server")
                                val url = dub.getString("url")
                                if (name == "MP4Upload") {
                                    if (dubServers.find { it.baseLink.contains(url.substringAfterLast("/")) } != null){
                                        continue
                                    }
                                }
                                if (name == "PDrain") {
                                    if (dubServers.find { it.baseLink.substringBeforeLast("?").contains(url.substringBeforeLast("?")) } != null){
                                        continue
                                    }
                                }
                                if (name == "1Fichier") {
                                    continue
                                }
                                val server = Server.check(activity, url)
                                if (dubServers.find { it.baseLink.substringAfterLast("/") == url.substringAfterLast("/") } == null) {
                                    dubServers.add(server?: WebServer(activity, url, name))
                                }
                            }
                        }
                    }
                }
                this@TVServersFactory.subServes = subServers.sortedWith(
                    compareBy(
                        { it.name.contains("(WEB)") },
                        {it.name}
                    )
                ).filter { it is WebServer || it.canStream }.toMutableList()
                this@TVServersFactory.dubServers = dubServers.sortedWith(
                    compareBy(
                        { it.name.contains("(WEB)") },
                        {it.name}
                    )
                ).filter { it is WebServer || it.canStream }.toMutableList()
                if (dubServers.isNotEmpty()) {
                    activity.startActivityForResult(
                        Intent(activity, TVMultiSelection::class.java),
                        REQUEST_CODE_MULTI
                    )
                } else {
                    this@TVServersFactory.servers = subServers.toMutableList()
                    showServerList()
                }
            } catch (e: Exception) {
                e.printStackTrace()
                FirebaseCrashlytics.getInstance().recordException(e)
                this@TVServersFactory.servers = ArrayList()
                serversInterface.onFinish(false, false)
            }
        }

    }

    interface ServersInterface {
        fun onReady(serversFactory: TVServersFactory)

        fun onFinish(started: Boolean, success: Boolean)
    }

    companion object {
        var REQUEST_CODE_LIST = 4456
        var REQUEST_CODE_OPTION = 6157
        var REQUEST_CODE_MULTI = 6497

        fun start(activity: FragmentActivity, url: String, chapter: RecentAV1, serversInterface: ServersInterface) {
            start(activity, url, chapter.name, chapter.chapter, chapter.asRecord(), null, serversInterface)
        }

        fun start(activity: FragmentActivity, url: String, name: String, chapter: String, record: Record, viewHolder: Presenter.ViewHolder?, serversInterface: ServersInterface?) {
            doAsync {
                serversInterface?.let {
                    val factory = TVServersFactory(activity, url, name, chapter, record, viewHolder, it)
                    serversInterface.onReady(factory)
                    factory.get()
                }
            }
        }
    }
}
